Vector collection documentation [HZAI-77] (#1124)
hazelcast/hz-docs: docs/modules/vector-search/pages/vector-collections.adoc

> = Vector Collection

Reviewer: I think we should put this under [Data structures](https://docs.hazelcast.com/hazelcast/5.4/data-structures/) - it fits there best, unless we want to promote it on top level

Author: 97e456281191372558f94773bdb743a9c1f92e4f

> 

Reviewer: ```suggestion
:page-enterprise: true
:page-beta: true

```

== Intro

> The primary object for interacting with the database is a Vector Collection. This collection holds information about the vectors and metadata (value).

Reviewer: I don't think we use "database" in context of vector collection, this is misleading.

The intro could contain more info about vector collection is used for (similarity search), what is the structure (key, value, vectors+indexes) - that is partially described.

Author: 97e456281191372558f94773bdb743a9c1f92e4f

A collection can consist of one or more indexes, all sharing a single storage metadata. Each index represents a distinct vector space. Vectors from different indexes are stored independently, and indexes can have different configurations.

NOTE: At this point, each metadata object must correspond to exactly one vector in each index.

> NOTE: When working with multiple indexes in your collections, it's necessary to specify the index to search in your queries..

Reviewer: this note is too specific for intro, it should be in section about search

Author: note deleted:
97e456281191372558f94773bdb743a9c1f92e4f

== Configuration

> Collection configuration can be set dynamically during vector collection creation or statically during cluster configuration. Below is a table listing all available options:

Reviewer: and, unlike other data structures, configuration has to exist before the collection can be used

Author: 64e5181268054046bf02460e682f405097c2dea6

.Collection configuration options
[cols="1,2,1,1",options="header"]
|===
|Option|Description|Required|Default

|name
|The name of the vector collection.
|Y
|null

|indexes
|Information about indexes configuration
|Y
|null
|===

.Index configuration options
[cols="1,2,1,1",options="header"]
|===
|Option|Description|Required|Default

|name
|The name of the vector index.
|`N` - for single-index vector collections. `Y` - for multi-index collection
|null

|dimension
|Vectors dimension
|Y
|null

|metric
|Distance metric used to calculate the distance between two vectors.

> See table Available distance metrics.

Reviewer 2: I know it's right next to this table, but perhaps best to replace with a link to `Available distance metrics`

Author: 6b265566fb67a94493a066d7015575e597f5ece5

|Y
|null

|max-degree
|The maximum number of neighbors per node is calculated as max-degree * 2
|Y
|16

|ef-construction
|The size of the search queue to use when finding nearest neighbors.
|Y
|100

|use-deduplication
|The option for vector deduplication.
When disabled, a new node will be created for each vector, even if it is the same as an existing one.
|true

|===

.Available distance metrics
[cols="2,2,2",options="header"]
|===
|Name|Description| Score definition

|EUCLIDEAN
|euclidean distance
|1 / (1 + squareDistance(v1, v2))

|COSINE
|cosine of the angle between the vectors
|(1 + cos(v1, v2)) / 2

|DOT
|dot product of the vectors
|(1 + dotProduct(v1, v2)) / 2

|===

> WARNING: The recommended method for computing cosine similarity is to normalize all vectors to unit length and utilize the DOT metric instead.

Author: Did not understand the remark about the table. The table is closed before this line.
warning -> note: 6b265566fb67a94493a066d7015575e597f5ece5

Configuration example:

[tabs]
====
XML::
+
--
[source,xml]
----
<hazelcast>
    <vector-collection name="vector-collection-name">
        <indexes>
            <index name="index-1">
                <dimension>2</dimension>
                <metric>DOT</metric>
            </index>
            <index name="index-2">
                <dimension>2</dimension>
                <metric>DOT</metric>
                <max-degree>10</max-degree>
                <ef-construction>20</ef-construction>

>                 <use-deduplication>false</use-deduplication>

Author: 6b265566fb67a94493a066d7015575e597f5ece5

            </index>
        </indexes>
    </vector-collection>
</hazelcast>
----
--
YAML::
+
--
[source,yaml]
----
hazelcast:
  vector-collection:
    vector-collection-name:
      indexes:
        - name: index-1
          dimension: 2
          metric: DOT
        - name: index-2
          dimension: 2
          metric: DOT
          max-degree: 10
          ef-construction: 20
          use-deduplication: false
----
--

> Java::

Reviewer 2: shall we also introduce a Python sample?

Author: The API for Python is different from that for Java. I expected it to be in a separate document.

+
--
[source,java]
----
Config config = new Config();
VectorCollectionConfig collectionConfig = new VectorCollectionConfig("vector-collection-name")
    .addVectorIndexConfig(
            new VectorIndexConfig()
                .setName("index-1")
                .setDimension(2)
                .setMetric(Metric.DOT)
    ).addVectorIndexConfig(
            new VectorIndexConfig()
                .setName("index-2")
                .setDimension(2)
                .setMetric(Metric.DOT)
                .setMaxDegree(10)
                .setEfConstruction(20)
                .setUseDeduplication(false)
    );
config.addVectorCollectionConfig(collectionConfig);
----
--
====

== Manage collection

=== Create collection

> To create a vector collection, use the static method `getCollection(HazelcastInstance instance, VectorCollectionConfig collectionConfig)` of the `VectorCollection` class.

Reviewer: `create a vector collection` / `get a vector collection` are not completely on the point. It does not matter if the collection exists. It matters if the _configuration_ exists. And you can use `getCollection(HazelcastInstance instance, VectorCollectionConfig collectionConfig)` many times provided that `collectionConfig` is always the same.

Author: 64e5181268054046bf02460e682f405097c2dea6

Example:

[tabs]
====
Java::
+
--
[source,java]
----
VectorCollection vectorCollection = VectorCollection.getCollection(hazelcastInstance, vectorCollectionConfig);
----
--
====

=== Get existing vector collection

To get a vector collection, use the static method `getCollection(HazelcastInstance instance, String collectionName)` of the `VectorCollection` class.

Example:

[tabs]
====
Java::
+
--
[source,java]
----
VectorCollection vectorCollection = VectorCollection.getCollection(hazelcastInstance, collectionName);
----
--
====

> === Optimize collection

Reviewer: I'd move optimization after more ordinary operations

Author: Optimize method is more part of managing the collection and not managing data.

Reviewer: but it is not the first thing you do after creating a collection. IMO more logical and easier to read order is: create/get collection, put/get data, search, optimize. Putting create and optimize together for the sake of grouping as "manage" does not justify a bit awkward reading order.

Author: 7e1b5c3452fd6c95ac9cd6d02014415617138c44

An optimization operation may be needed in the following cases:

* To permanently delete nodes from the graph that were marked for removal.
* After adding a significant number of vectors.
* If some nodes become unreachable and the search results return fewer vectors than expected.

> NOTE: The optimization operation is time-consuming process, and no mutable operations are allowed during this process.

Reviewer: ```suggestion
NOTE: The optimization operation is a time-consuming process, and no mutating operations are allowed during this process.
```

[tabs]
====
Java::
+
--
[source,java]
----
VectorCollection vectorCollection = VectorCollection.getCollection(hazelcastInstance, collectionName);
CompletionStage<Void> result = vectorCollection.optimizeAsync("index_name");
----
--
====

== Manage data

> All methods of `VectorCollection` that work with collection data are asynchronous. The result is returned as a `CompletionStage`. A collection interacts with objects in the form of documents (`VectorDocument`). Each document comprises a value and one or more vectors associated with that value.

Reviewer: I think we do not need to describe what is obvious from Java API (async, CompletionStage) but mention about the role of `VectorDocument` is important

Reviewer: but in context of what @EugeneAbramchuk asked in slack, it may be good to mention that concurrency must be controlled when async API is used, so users do not shoot themselves in the foot. BTW, Jet gives that out of the box

Author: Do you mean this one? https://hazelcast.slack.com/archives/C06RQRE2MEV/p1717604096835489.
I am not entity understand what you suggest. Do you suggest including information in the documentation that asynchronous operations may incur additional heap overhead?

Reviewer: I am not sure what kind heap problem @vbekiaris described, it looks like related to another discussion not to vector search.

with async API it is easy to send many (too many) requests, and each of them will consume heap on server (client also) during processing. but if the client controls the concurrency (relying on clients doing things correctly is of course risky...), then everything should run fine and I am not aware of any other problems specific to async API which do occur also for sync API.

Reviewer 3: > I am not sure what kind heap problem @vbekiaris described, it looks like related to another discussion not to vector search.

right, my note was about JVM heap overheads per `IMap` per partition (that was the topic of the earlier discussion Eugene linked to), unrelated to vector search.

> with async API it is easy to send many (too many) requests, and each of them will consume heap on server (client also) during processing.

That's the issue that's relevant to describe in relation to async requests.

Author: Hope you mean to add something like that:
b21e6929889ee3ffe43e883f5baf927a4c76bd29

=== Create document
To create a document use statics the methods of `VectorDocument` and `VectorValues` classes.

Example single index vector collection:
[tabs]
====
Java::
+
--
[source,java]
----
VectorDocument<String> document = VectorDocument.of(
        "document value",
        VectorValues.of(
                new float[]{0.2f, 0.9f, -1.2f}
        )
);
----
--
====

For multi-index collections, specify the names of the indexes to which the vectors belong:
[tabs]
====
Java::
+
--
[source,java]
----
VectorDocument<String> document = VectorDocument.of(
        "some document value",
        VectorValues.of(
                Map.of(
                        "index1", new float[] {1.2f, -0.3f, 2.2f},
                        "index2", new float[] {0.2f, 1.3f, -0.2f}
                )
        )
);
----
--
====

> === Put objects

Reviewer: are we calling them "objects" or "entries"? I prefer entries. But we use also VectorDocument and value 

entry = key + VectorDocument

Author: b2fd6c714396ed311c966efd372253f1d485ff18

To put single document to a vector collection, use the method `putAsync`, `putIfAbsent` or `setAsync` of the `VectorCollection` class.
[tabs]
====
Java::
+
--
[source,java]
----
VectorCollection<String, String> vectorCollection = VectorCollection.getCollection(hazelcastInstance, collectionName);
VectorDocument<String> document = VectorDocument.of("document value", VectorValues.of(new float[] {0.2f, 0.9f, -1.2f}));

> CompletionStage<VectorDocument<String>> result = vectorCollection.putAsync("1", document);

Reviewer: maybe the examples should show waiting for the result (sync invocation)?

Author: I believe dealing with CompletionStage is beyond the scope of this example. Hopefully, the user is already familiar with how to use it.

----
--
====

To put several documents to a vector collection, use the method `putAllAsync` of the `VectorCollection` class.
[tabs]
====
Java::
+
--
[source,java]
----
VectorCollection<String, String> vectorCollection = VectorCollection.getCollection(hazelcastInstance, collectionName);
VectorDocument<String> document1 = VectorDocument.of("some document1 value", VectorValues.of(new float[] {0.2f, 0.9f, -1.2f}));
VectorDocument<String> document2 = VectorDocument.of("some document2 value", VectorValues.of(new float[] {1.2f, -0.3f, 2.2f}));
CompletionStage<Void> result = vectorCollection.putAllAsync(Map.of("1", document1, "2", document2));
----
--
====

=== Read objects
To get document from a vector collection, use the method `getAsync` of the `VectorCollection` class.

[tabs]
====
Java::
+
--
[source,java]
----
VectorCollection<String, String> vectorCollection = VectorCollection.getCollection(hazelcastInstance, collectionName);
CompletionStage<VectorDocument<String>> result = vectorCollection.getAsync("1");
----
--
====

=== Update objects

> To update single document in a vector collection, use the method `putAsync` or `setAsync` of the `VectorCollection` class.

Reviewer: "Update objects", "update single document" - we should use consistent names

Author: b2fd6c714396ed311c966efd372253f1d485ff18

[tabs]
====
Java::
+
--
[source,java]
----
VectorCollection<String, String> vectorCollection = VectorCollection.getCollection(hazelcastInstance, collectionName);
VectorDocument<String> document = VectorDocument.of("some document value", VectorValues.of(new float[] {0.2f, 0.9f, -1.2f}));
CompletionStage<Void> result = vectorCollection.setAsync("1", document);
----
--
====

=== Delete objects
To delete document from a vector collection, use the method `deleteAsync` or `removeAsync` of the `VectorCollection` class.

[tabs]
====
Java::
+
--
[source,java]
----
VectorCollection<String, String> vectorCollection = VectorCollection.getCollection(hazelcastInstance, collectionName);
CompletionStage<Void> resultDelete = vectorCollection.deleteAsync("1");
CompletionStage<VectorDocument<String>> resultRemove = vectorCollection.removeAsync("2");
----
--
====

> NOTE: These methods do not delete index nodes but mark them as deleted. This may impact search speed and memory usage. To permanently remove nodes from the index, you must run index optimization after deletion. See <<optimize-collection, optimize method>>

Reviewer 2: Also marked as a table element, similar to above

Author: this is a preview issue in github

== Similarity search

Vector search returns objects with vectors that are most similar to that of the query, based on specified metrics.

> When submitting a search query, you have the option to include one vector values representing your query embeddings and specify a limit on the number of results to retrieve.

Reviewer: > include one vector values representing your query embeddings

this is hard to understand, but it's hard to describe `isIncludeVectors` concisely

Author: b2fd6c714396ed311c966efd372253f1d485ff18

=== Single-index vector collection search

For similarity search, use the `searchAsync` method of the `VectorCollection`.
[tabs]
====
Java::
+
--
[source,java]
----
CompletionStage<SearchResults<String, String>> results = vectorCollection.searchAsync(
        VectorValues.of(new float[] {0, 0}),
        SearchOptions.builder()
            .limit(10)
            .includeVectors()
            .includeValue()
            .build()
);
----
--
====

> === Multi-index vector collection search

Reviewer: It is not obvious if "Multi-index" applies to "vector collecton" or "search". Is it worth distingushing those cases so much? Maybe just describe in text, that if the collection has 1 index, you do not have to specify the name, but when it has more, then you have to.

Author: b2fd6c714396ed311c966efd372253f1d485ff18

For similarity search, use the `searchAsync` method of the `VectorCollection`, specifying the index name to search.
[tabs]
====
Java::
+
--
[source,java]
----
CompletionStage<SearchResults<String, String>> results = vectorCollection.searchAsync(
        VectorValues.of("index_name", new float[] {0, 0}),
        SearchOptions.builder()
            .limit(10)
            .includeVectors()
            .includeValue()
            .build()
);
----
--
====

=== Similarity search options
Search parameters are passed as an argument `searchOptions` to the searchAsync method.

> .Search options

Reviewer 2: a line break is missing before `.Search options`

Author: 6b265566fb67a94493a066d7015575e597f5ece5

[cols="1,2,3a",options="header"]
|===
|Option|Description|Example

|limit

> |How many vectors should be returned in search result

Reviewer: ```suggestion
|How many results should be returned
```

|

[tabs]
====
Java::
+
--
[source,java]
----

> var options = SearchOptions.builder().limit(10).build();

Reviewer: default limit is currently 1, but it is not mentioned in the Javadoc (do we want to commit to that?)

Author: b2fd6c714396ed311c966efd372253f1d485ff18

----
--
====

|includeValue
|Include the user value in the search result.
|

[tabs]
====
Java::
+
--
[source,java]
----

> var options = SearchOptions.builder().includeValue().build();

Reviewer: these partial examples are misleading and rely on default limit = 1. Do we really want to provide such trivial code? Naming method + 1 example with everything should be sufficient

----
--
====

|includeVectors
|Include the vectors values in the search result.
|

[tabs]
====
Java::
+
--
[source,java]
----
var options = SearchOptions.builder().includeVectors().build();
----
--
====

|hints
|Extra hints for the search.
|

[tabs]
====
Java::
+
--
[source,java]
----

> var options = SearchOptions.builder().hint(Hints.PARTITION_LIMIT, 1).build();

Reviewer: hints are tricky and subject to change. We need to have a clear messaging here about them.

Author: What exactly details do you want to mention about the hints?

Reviewer: add the following disclaimer:

```
WARNING: hints allow fine-tuning some aspects of search execution but are subject to change and may be removed in future versions.
```

Author: a1ea853e7d1f346965b2aea139b03c13410fee13
